### ISSUE TITLE
SphericalPolygonSet(double, S2Point...) fails for high resolution boundaries

### ISSUE THREAD

**Issue by wardev** [Hipparchus-Math/hipparchus#46]
Use case:
1. Use external program to design boundary
2. Export high resolution boundary (closely spaced points)
3. Import in Hipparchus
4. Use `hyperplaneThickness` to control performance/fidelity tradeoff

Below is a failing test case adapted from the ZigZag test case. The ZigZag region is enlarged because this issue does not occur for very small regions. The resulting region is ~ 30 degrees across. The ZigZag boundary is then sub-sampled to a tenth of the tolerance. The sub sampled data is what would be read in from an external file in the above use case. The sub-sampled data is used to create a new SPS, which is tested. On my machine a NPE is generated by the call to `getEnclosingCap()`. 

I've also included a plot of the region. Blue dots are the sub-sampled points, black lines and dots are the computed boundary. It matches closely in some locations, others are very far off.

@maisonobe any help or pointers would be appreciated. I'm still trying to get my head around the BSPTree and partitioning code.

```java
    @Test
    public void testZigZagBoundary() {
        final double tol = 1.0e-4;
        // sample region, non-convex, not too big, not too small
        final S2Point[] vertices = {
                new S2Point(-0.12630940610562444e1, (0.8998192093789258 - 0.89) * 100),
                new S2Point(-0.12731320182988207e1, (0.8963735568774486 - 0.89) * 100),
                new S2Point(-0.1351107624622557e1, (0.8978258663483273 - 0.89) * 100),
                new S2Point(-0.13545331405131725e1, (0.8966781238246179 - 0.89) * 100),
                new S2Point(-0.14324883017454967e1, (0.8981309629283796 - 0.89) * 100),
                new S2Point(-0.14359875625524995e1, (0.896983965573036 - 0.89) * 100),
                new S2Point(-0.14749650541159384e1, (0.8977109994666864 - 0.89) * 100),
                new S2Point(-0.14785037758231825e1, (0.8965644005442432 - 0.89) * 100),
                new S2Point(-0.15369807257448784e1, (0.8976550608135502 - 0.89) * 100),
                new S2Point(-0.1526225554339386e1, (0.9010934265410458 - 0.89) * 100),
                new S2Point(-0.14679028466684121e1, (0.9000043396997698 - 0.89) * 100),
                new S2Point(-0.14643807494172612e1, (0.9011511073761742 - 0.89) * 100),
                new S2Point(-0.1386609051963748e1, (0.8996991539048602 - 0.89) * 100),
                new S2Point(-0.13831601655974668e1, (0.9008466623902937 - 0.89) * 100),
                new S2Point(-0.1305365419828323e1, (0.8993961857946309 - 0.89) * 100),
                new S2Point(-0.1301989630405964e1, (0.9005444294061787 - 0.89) * 100)};
        SphericalPolygonsSet zone = new SphericalPolygonsSet(tol, vertices);
        // sample high resolution boundary
        List<S2Point> points = new ArrayList<>();
        final Vertex start = zone.getBoundaryLoops().get(0);
        Vertex v = start;
        double step = tol / 10;
        do {
            Edge outgoing = v.getOutgoing();
            final double length = outgoing.getLength();
            int n = (int) (length / step);
            for (int i = 0; i < n; i++) {
                points.add(new S2Point(outgoing.getPointAt(i*step)));
            }
            v = outgoing.getEnd();
        } while (v != start);
        // create zone from high resolution boundary
        zone = new SphericalPolygonsSet(tol, points.toArray(new S2Point[0]));
        //print(zone);
        EnclosingBall<Sphere2D, S2Point> cap = zone.getEnclosingCap();
        // check cap size is reasonable. The region is ~0.5 accross, could be < 0.25
        Assert.assertTrue(cap.getRadius() < 0.5);
        for (S2Point vertex : vertices) {
            // check original points are on the boundary
            Assert.assertEquals("" + vertex, Location.BOUNDARY, zone.checkPoint(vertex));
            // check original points are within the cap
            Assert.assertTrue("" + vertex, cap.contains(vertex));
        }
    }
```

![image](https://user-images.githubusercontent.com/5224597/45381169-93aabd00-b5d2-11e8-93dd-c830966ed869.png)

**Comment by maisonobe** [comment]
From the plot, I would say that it is probably a numerical issue due to intersection of too close circles.
If you have one regular segment of the boundary (i.e. one small arc of circle) that is split into several
parts connected by intermediate vertices, then when building the region back from these vertices
implies computing the intersections of the arcs that join them. However, here all the arcs belong to
almost the same circle, the differences being only in the numerical noise when building the vertices.

This is *very* difficult to solve.

There is a comment in the constructor you used that explicitly warn against such cases:

     Polygons with thin pikes or dents are inherently difficult to handle because
     they involve circles with almost opposite directions at some vertices. Polygons
     whose vertices come from some physical measurement with noise are also
     difficult because an edge that should be straight may be broken in lots of
     different pieces with almost equal directions. In both cases, computing the
     circles intersections is not numerically robust due to the almost 0 or almost
     π angle. Such cases need to carefully adjust the hyperplaneThickness
     parameter. A too small value would often lead to completely wrong polygons
     with large area wrongly identified as inside or outside. Large values are
     often much safer. As a rule of thumb, a value slightly below the size of the
     most accurate detail needed is a good value for the hyperplaneThickness
     parameter.

**Comment by maisonobe** [comment]
Perhaps one thing we could do is provide a filtering capability to decimate the initial array before
it is used, by just comparing the angles between incoming and outgoing arcs at one vertex. If the
angle is too close to π, then the point would be considered spurious and ignored.

**Comment by wardev** [comment]
That makes sense. I tried playing around with the tolerance, but I couldn't find a value that produced something reasonable. I'll try implementing your suggestion. Thanks for the tips!

**Comment by maisonobe** [comment]
It would also be worth porting this to Euclidean/2D once it is available. The same issues exist in
Euclidean geometry, and the same comment is in the constructor.

**Comment by maisonobe** [comment]
In fact, rather than checking the angle between the arcs, I think it would be more straightforward to use only the points themselves. we consider points P0, P1 and P2. We build the circle (spherical geometry)/line (Euclidean geometry) from P0 to P2. If the distance between P1 and this circle/line is closer than the tolerance, we drop P1.

**Comment by wardev** [comment]
I implemented decimation using the criteria in your last comment. Plot below. Blue points are those extracted by decimation algorithm, black points are those on the resulting SPS boundary. It seems that two extra points are introduced on the longer edges. I'm going to look in to how the extra points are added and improving the performance.
![image](https://user-images.githubusercontent.com/5224597/45430670-dd97af80-b673-11e8-89d0-d9ee14ec52e0.png)

**Comment by maisonobe** [comment]
This is a huge improvement!
Do all the existing tests pass (including Orekit tests)?

**Comment by wardev** [comment]
All tests pass except for `testGitHubIssue42B`, which is sensitive to the order of the points in the input array. Without any modified code just starting the boundary from a different point will cause this test to fail with an exception saying the boundary is open. I think this is a separate issue since the created region should be independent of which boundary point is first on the list. At first glance it looks like this should be an easy region because it is a small square and the edges are much longer than ulp(1). Why is this region hard?

Do you have a tool to plot the output of `RegionDumper.dump()` so I can look at it visually?

I'm also noticing significant performance improvements. If I understand the existing algorithm correctly it is O(n^2), where n is the length of the input array. With decimation it is O(m^2 + n lg n), where m is the number of edges need to define the region.

**Comment by maisonobe** [comment]
> All tests pass except for `testGitHubIssue42B`, which is sensitive to the order of the points in the input array. Without any modified code just starting the boundary from a different point will cause this test to fail with an exception saying the boundary is open. I think this is a separate issue since the created region should be independent of which boundary point is first on the list. At first glance it looks like this should be an easy region because it is a small square and the edges are much longer than ulp(1). Why is this region hard?

With the implementation before the fix, one of the intersection points was computed differently in two different parts of the code and due to very small tolerances, the two results were considered to be different points. The fix for issue 42 was to remember some topological information (mainly what hyperplanes were involved in the computation of the point) so the point was computed only once and properly identified. So the region itself is an easy one, but setting a tolerance just at the level of numerical noise was a problem.

> 
> Do you have a tool to plot the output of `RegionDumper.dump()` so I can look at it visually?

No, I use this mainly following the depth in the tree with the result open in a text file, in order
to understand where I am when dealing with the parent and plus/minus children in the debugger.

> 
> I'm also noticing significant performance improvements. If I understand the existing algorithm correctly it is O(n^2), where n is the length of the input array. With decimation it is O(m^2 + n lg n), where m is the number of edges need to define the region.

This is a good thing.

We can therefore close the issue when you push this to master (this is usually done automatically
in GitHub by using special comments in the commit message).

**Comment by maisonobe** [comment]
Do you have something to add or could you push your fix to master?

**Comment by wardev** [comment]
Sorry for the delay. I was tied up with end of FY stuff. In its current state the main code is working, but I need to finish writing unit tests and testing it. If you want it now I can push it up, though it doesn't meet the quality standards. If you're OK waiting a couple weeks I can clean it up and finish the testing.

I did notice another issue while I was doing the testing and I was planning to open a separate tracker for it. In brief the docs for the enclosing cap state that any point external to the cap is external to the region, which is not strictly true. There may be points on the boundary of the region arbitrarily far from cap. I think the docs for enclosing cap should be changed to state that any point internal to the region is internal to the cap. Otherwise there will be more code to fix.

I think the cause of the issue is that tolerance for regions is defined as a plane thickness, but tolerance for caps is defined as an extra distance from the center of the cap on the surface of the sphere. So when two planes intersect with a small angle between them the furthest point of the intersection of the slabs (thick planes) can be far from the intersection of the planes. In fact that distance becomes infinite as the angle between the planes vanishes.

**Comment by maisonobe** [comment]
I am fine with a few weeks delay.
A separate issue would be better for the enclosing cap, and documentation update seems OK. Geometry on the sphere is anyway tricky, so trying to make it intuitive is a dead end. Just try to document what we could achieve rather that what we would have expected.

Thanks a lot!

**Comment by wardev** [comment]
Alright, the changes are ready to push. I think tests in EllipsoidTessellatorTest are sensitive to the order in which the S2Points are passed to SphericalPolygonSet, perhaps because it changes the starting point for the tessellation algorithm. So I made a change the the Hipparchus code to preserve the order of the points when possible for better backward compatibility. The tolerances on EllipsoidTessellatorTest probably could be loosened though.

**Comment by maisonobe** [comment]
Thanks Evan!
For tolerances in EllipsoidTessellatorTest, this should be discussed on the Orekit forge ;-)
